### DISCUSSION TITLE
Jinja2 export template generates incorrectly value 'core.ObjectType.None' of field object_types on extras.custom-fields model

### DISCUSSION THREAD

**Discussion by marcusyuri** [netbox-community/netbox#17687] [Help Wanted!] [answered]
### Deployment Type

Self-hosted

### NetBox Version

v4.1.3

### Python Version

3.10

### Steps to Reproduce

1. Create a custom field and assign to an object type. Ex: my_custom_field assigned to dcim.interface
![image](https://github.com/user-attachments/assets/820e5499-b3b6-4492-8927-4de7d408eb94)

3. Create a Jinja2 export template on Custom Field matching field object_types
```
{% for field in queryset %}
name: {{ field.name }}
object_types: {{ field.object_types }}
{% endfor %}
```
5. Export the custom field using the export template
![image](https://github.com/user-attachments/assets/29ff97e9-7cdf-4655-a95c-41e82a9750e5)

### Expected Behavior

The object_types field replaced by the assigned object type. In the example, it was expected:

object_types: dcim.interface

### Observed Behavior

object_types: core.ObjectType.None

**Comment by jeremystretch** [comment]
`object_types` is a manager that can return multiple ContentType instances. It needs to be invoked as a queryset, e.g. `field.object_types.all()`.

**Reply by marcusyuri** [reply]
Hello, Jeremy!

Thanks for the answer. I changed my template to:

```
{% for field in queryset %}name: {{ field.name }}
type: {{ field.type }}
search_weight: {{ field.search_weight }}
filter_logic: {{ field.filter_logic }}
weight: {{ field.weight }}
object_types:{% for type in field.object_types.all() %}
- {{ type.app_label }}.{{ type.model }}{% endfor %}
{{ "---
" if not loop.last }}{% endfor %}
```

And I can successfull generate an YAML export file:
```
name: PortLocation
type: object
search_weight: 1000
filter_logic: loose
weight: 100
object_types:
- dcim.frontport
- dcim.rearport
---
name: cust_id
type: text
search_weight: 1000
filter_logic: exact
weight: 100
object_types:
- tenancy.tenant
---
name: custom_test
type: integer
search_weight: 1000
filter_logic: loose
weight: 100
object_types:
- dcim.site
---
name: custom_test1
type: text
search_weight: 1000
filter_logic: loose
weight: 100
object_types:
- dcim.device
---
name: custom_test2
type: integer
search_weight: 1000
filter_logic: loose
weight: 100
object_types:
- dcim.device
```

I have an additional question: What is the expected format to import, as YAML, the object_types list??? If I try to import the generated YAML, I get the following error:
![image](https://github.com/user-attachments/assets/87853b94-6fab-43ef-a47e-7716b8a094dd)

Also, if I put the app_label I get the same error (missing attribute app_label):
```
name: custom_test
type: integer
search_weight: 1000
filter_logic: loose
weight: 100
object_types:
- app_label: dcim
  model: site
```

Can you help me????

**Reply by candlerb** [reply]
> If I try to import the generated YAML, I get the following error:

You cannot use exported data as a guide to the import format. They are incompatible with each other by design, and there are no plans to change this (see [FAQ](https://github.com/netbox-community/netbox/wiki/Frequently-Asked-Questions#why-cant-i-pass-exported-data-directly-into-the-import-form))

I suggest you approach this the other way round: show us the YAML you tried to import, and we can suggest alternatives. I am guessing you tried something like this:

```
name: custom_test
type: integer
search_weight: 1000
filter_logic: loose
weight: 100
object_types:
- dcim.site
```

which does give that error. There is code in `netbox/extras/forms/bulk_import.py`:

```
class CustomFieldImportForm(CSVModelForm):
    object_types = CSVMultipleContentTypeField(
        label=_('Object types'),
        queryset=ObjectType.objects.with_feature('custom_fields'),
        help_text=_("One or more assigned object types")
    )
```

and `netbox/utilities/forms/fields/csv.py`

```
class CSVMultipleContentTypeField(forms.ModelMultipleChoiceField):
    """
    CSV field for referencing one or more content types, in the form `<app>.<model>`.
    """
    STATIC_CHOICES = True

    # TODO: Improve validation of selected ContentTypes
    def prepare_value(self, value):
        if type(value) is str:
            ct_filter = Q()
            for name in value.split(','):
                app_label, model = name.split('.')
                ct_filter |= Q(app_label=app_label, model=model)
            return list(ContentType.objects.filter(ct_filter).values_list('pk', flat=True))
        return object_type_identifier(value)
```

Now, it looks to me from that like it expects a string which is a comma-separated list of object types. So I tried this:

```
name: custom_test
type: integer
search_weight: 1000
filter_logic: loose
weight: 100
object_types: dcim.site
```

and it works for me. YMMV.

P.S. Since there's an explicit test for `if type(value) is str` then it seems reasonable to me that it could also have a case `if type(value) is list`. and be able to import the YAML list form. You could raise that as a feature request I guess, but there are probably many other cases like this, essentially aimed at a single CSV column which could contain multiple values.

**Reply by marcusyuri** [reply, marked as answer]
Hello, Brian!

          Thanks for the response. My goal is to export the data from custom_field model in a format that can be imported without any changes or processing in netbox.

          I just changed my export template to generate a single line, comma separated, object_types:

```
{% for field in queryset %}name: {{ field.name }}
type: {{ field.type }}
search_weight: {{ field.search_weight }}
filter_logic: {{ field.filter_logic }}
weight: {{ field.weight }}
object_types: {% for type in field.object_types.all() %}{{ type.app_label }}.{{ type.model }}{{ "," if not loop.last }}{% endfor %}
{{ "---
" if not loop.last }}{% endfor %}
```
The result is the following:
```
name: SAP_Nummer
type: text
search_weight: 1000
filter_logic: loose
weight: 100
object_types: dcim.device,dcim.devicebay,dcim.devicerole,dcim.devicetype
---
name: code_ope
type: text
search_weight: 1000
filter_logic: loose
weight: 100
object_types: circuits.circuit
---
name: cust_id
type: text
search_weight: 1000
filter_logic: exact
weight: 100
object_types: tenancy.tenant
```
This appears to solve the problem!!!!
